sdk/core/Azure.Core/src/ResponseExceptionFactory.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Azure.Core.Pipeline
+{
+    internal class ResponseExceptionFactory
```

Comment:
This is essentially what's in ClientDiagnostics, just one approach to factoring out the logic (that I'm not committed to if we don't like it)